### PR TITLE
[Kernel] Add GPU kernels and enable LLaMA model.

### PR DESCRIPTION
**Pull request by changqi1**
# How to build and test w/ single Intel GPU

```bash
source /opt/intel/oneapi/setvars.sh
source ../3rdparty/oneccl/build/_install/env/setvars.sh

export CC=icx
export CXX=icpx
mkdir build && cd build
cmake .. -DWITH_GPU=ON
make -j

OMP_NUM_THREADS=20 mpirun -n 1 -env XFT_ENGINE=GPU:0 numactl -N 0 -m 0 ./example \
        --model /home/xfast/models/llama-2-7b-chat-xft/ \
        --token /home/xfast/models/llama-2-7b-chat-hf/tokenizer.model \
        --dtype fp16 \
        --loop 3 \
        --no_stream \
        --input_len 18 \
        --output_len 8

[INFO] First token time: 69.575 ms
[INFO] Second token time: 61.2469 ms
[INFO] Final output is:
==============================================
Once upon a time, there existed a little girl who liked to have adventures. She lived in a small village surrounded by
```

### CODE REVIEW

File: src/models/model_factory.h
Comment (pujiang2018) [inline src/models/model_factory.h:1]: Do not need to change this file.

File: src/utils/matmul_helper.h [hunk 1]
```diff
@@ -349,8 +349,11 @@ class MMHelper {
         // W8A8
         else if constexpr (std::is_same_v<WeiT, w8a8_t>) {
             using dt = dnnl::memory::data_type;
+            dnnl::engine eng(dnnl::engine::kind::cpu, 0);
+            dnnl::stream stm(eng);
```
Comment (pujiang2018) [inline src/utils/matmul_helper.h:353]: why now need to create an engine and stream every time calling into the function? it may impact the performance.

Comment (changqi1) [reply src/utils/matmul_helper.h]: Use unified onednn engine and stream.

File: src/layers/mlp_llama.h [hunk 2]
```diff
@@ -275,8 +275,7 @@ class LlamaMLP : public SingletonBase<LlamaMLP<WeiT>> {
         }
     }
 
-    template <typename T1, typename T2>
-    void catGateUpProj(DecoderContext *ctx, hpj::Matrix<T1> &input, hpj::Matrix<T2> &output, hpj::Matrix<T2> &siluBuf) {
+    void catGateUpProj(DecoderContext *ctx, hpj::Matrix<InT> &input, hpj::Matrix<ImT> &output, hpj::Matrix<ImT> &siluBuf) {
```
[the remark this answers is not in the document]
Comment (changqi1) [reply src/layers/mlp_llama.h]: No Change

File: src/models/model_factory.h
Comment (changqi1) [reply src/models/model_factory.h:1]: compiler warning.

File: src/layers/attention.h [hunk 3]
```diff
@@ -294,6 +294,11 @@ class Attention {
                 std::iota(posIds.begin(), posIds.end(), pastSeqLen);
             }
             qkpo.forward(query.Data(), key.Data(), query.Stride(), key.Stride(), qkShape, posIds.data());
+#ifdef GPU
+            sycl::queue *q = static_cast<sycl::queue *>(ctx->device);
+            int64_t size = ctx->batchSize * ctx->inputSeqLen * qkvCols * sizeof(float);
+            q->memcpy(qkvMatMul.Data(), query.Data(), size).wait();
```
[the remark this answers is not in the document]
Comment (changqi1) [reply src/layers/attention.h]: need to copy attention data from GPU memory.

File: src/utils/matmul_helper.h [hunk 4]
```diff
@@ -1398,18 +1436,19 @@ class MMHelper {
...
     };
 
-    std::string create_key(bool transA, int M, int N, int K, int matmul_kind) {
-        std::string key = std::to_string(transA) + "_" + std::to_string(M) + "_" + std::to_string(N) + "_"
-                + std::to_string(K) + "_" + std::to_string(matmul_kind);
-        return key;
+    template <typename Twei>
+    std::string create_key(bool transA, int M, int N, int K, int matmul_kind, const Twei *packedB) {
+        std::stringstream key;
+        key << transA << "_" << M << "_" << N << "_" << K << "_" << matmul_kind << "_" << packedB;
```
Comment (pujiang2018) [inline src/utils/matmul_helper.h:1442]: next step we need a better version for this function, OK with current version.

File: src/utils/matmul_helper.h [hunk 5]
```diff
@@ -1420,33 +1459,273 @@ class MMHelper {
...
+                }
+                default: {
+                    printf(">>> onednn_gemm_compute: postAlg type %s not supported.", std::to_string(postAlg).c_str());
+                    exit(-1);
+                }
+            }
+
+            if (postAlg == matmul_kinds::Basic) {
+                if (bias != nullptr)
+                    matmul_pd = new matmul::primitive_desc(*engine, input_md, weight_md, bias_md, output_md);
```
Comment (pujiang2018) [inline src/utils/matmul_helper.h:1647]: deleted the object or not?

File: tests/ut/attention_kernels_test.cpp [hunk 6]
```diff
@@ -86,7 +86,7 @@ static void selfAttentionRef(bfloat16_t *output, bfloat16_t *query, bfloat16_t *
         int kvHeadNum, int headSize, int oStride, int qStride, int kvStride, int batchSize, const int *tokenSizes,
         const float scale) {
 
-    int rowOffsets[batchSize] = {0};
+    int rowOffsets[batchSize];
```
[the remark this answers is not in the document]
Comment (changqi1) [reply tests/ut/attention_kernels_test.cpp:89]: Add memset. If use the style, the ICX will report error.

File: src/utils/matmul_helper.h (hunk 5 above)
Comment (changqi1) [reply src/utils/matmul_helper.h:1659]: delete in destructor

File: src/layers/rms_norm.cpp [hunk 7]
```diff
@@ -29,21 +34,58 @@ RmsNorm::RmsNorm() {
...
+}
+
+void RmsNorm::forward(const float *input, bfloat16_t *output, int rows, int iStride, int oStride, float epsilon) {
+    TimeLine t("RmsNorm.forward");
+}
+
+void RmsNorm::forward(const bfloat16_t *input, bfloat16_t *output, int rows, int iStride, int oStride, float epsilon) {
+    TimeLine t("RmsNorm.forward");
+    sycl::queue *gpu_queue = static_cast<sycl::queue *>(device);
+    // fastertransformer::invokeGeneralT5LayerNorm(
```
[the remark this answers is not in the document]
Comment (changqi1) [reply src/layers/rms_norm.cpp]: Done

File: src/layers/mlp_llama.h [hunk 8]
```diff
@@ -38,12 +38,10 @@
...
 // But please also be noted: we extended the MLP to include layer norm
-template <typename WeiT, typename InT = float, typename ImT = float, typename OutT = float>
-class LlamaMLP : public SingletonBase<LlamaMLP<WeiT>> {
+template <typename WeiT, typename InT = float, typename ImT = float, typename OutT = float, typename NORM_CLS = xft::RmsNorm>
+class LlamaMLP {
 public:
-    LlamaMLP() {}
-
-    LlamaMLP(DecoderContext *ctx) {}
+    LlamaMLP(DecoderContext *ctx) { norm = new NORM_CLS(ctx); }
```
[the remark this answers is not in the document]
Comment (changqi1) [reply src/layers/mlp_llama.h]: using non-pointer objects.

File: src/layers/layer_norm.cpp [hunk 9]
```diff
@@ -52,11 +59,18 @@ void LayerNorm::setWeight(const std::string &gammaPath, const std::string &betaP
 
 // input and output are in shape of (rows, normSize)
 // TODO: column-wise parallel
+#ifdef GPU
+void LayerNorm::forward(const float *input, float *output, int rows, int iStride, int oStride, float epsilon) {
```
[the remark this answers is not in the document]
Comment (changqi1) [reply src/layers/layer_norm.cpp:63]: Done

File: src/layers/attention.h [hunk 10]
```diff
@@ -51,6 +51,8 @@ class Attention {
         //todo(marvin): clear this code after all rotary_emb refactor
         if constexpr (std::is_same<QKPO_CLS, LlamaRotaryEmbedding>::value) { qkpo = LlamaRotaryEmbedding(ctx); }
 
+        norm = new NORM_CLS(ctx);
```
[the remark this answers is not in the document]
Comment (changqi1) [reply src/layers/attention.h]: using non-pointer objects.

File: src/common/sequence.h [hunk 11]
```diff
@@ -81,6 +82,20 @@ class SequenceMeta {
         , promptTokens(_inputSeqLen, 0)
         , step(0) {}
 
+    SequenceMeta(int32_t _sequenceID, std::vector<int32_t> &_promptTokens)
```
[the remark this answers is not in the document]
Comment (changqi1) [reply src/common/sequence.h]: Done

**Review by pujiang2018 — APPROVED** [review]

**Pull request merged**